feat/async_v2: revamping Asynchronous Programming section (#36)
fsprojects/fsharp-cheatsheet: docs/fsharp-cheatsheet.md

In contrast to .NET Tasks, async expressions are "cold" (need to be explicitly started) and every execution [propagates a CancellationToken implicitly](#asynchronous-programming-cancellation-async).

    open System
    open System.Threading
    open System.IO

    let readFile filename = async {
        do! Async.Sleep(TimeSpan.FromSeconds 5)  // use do! when awaiting an Async
        let! text = File.ReadAllTextAsync(filename) |> Async.AwaitTask  // (1)
        printfn "Finished Reading File"
        return text
    }

    // compose a new async computation from exising async computations
    let readFiles = [ readFile "A"; readFile "B" ] |> Async.Parallel

    // execute async computation
    let textOfFiles: string[] = readFiles |> Async.RunSynchronously
    // Out: Finished Reading File
    // Out: Finished Reading File

    // re-execute async computation again
    let textOfFiles': string[] = readFiles |> Async.RunSynchronously
    // Out: Finished Reading File
    // Out: Finished Reading File

> (1) As .NET Tasks became the central component of task-based asynchronous programming after F# Async were introduced, F#'s Async has `Async.AwaitTask` for interoperability.

Reviewer: ```suggestion
(1) While `task {}` expressions can invoke an `Async` directly, there's an explicit `Async.AwaitTask` that's used to map from `Task<'T>` to `Async<'T>`. Note there are some [known issues}https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/141) that are [not yet resolved in `FSharp.Core`](https://github.com/fsharp/fslang-suggestions/issues/840), so some codebases use an `AwaitTaskCorrect` helper instead.
```

Author: I reworded this a little bit so not to overwhelm the reader. The cancellation section already addresses the issue, and I'll have an exception handling section to mention the other piece.

#### Creation / Composition

The `Async` module has a number of functions to compose and start computations. The full list with explanations can be found in the [Async Type Reference](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-control-fsharpasync.html#section0).

| Function                | Description                                                                                                                                                                                                                     |
|-------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Async.Ignore            | Creates an `Async<unit>` computation from an `Async<'T>`                                                                                                                                                                        |

> | Async.Parallel          | Composes a new computation from multiple computations, `Async<'T> seq`, and runs them in parallel; it returns all the results in an array `Async<'T array>`                                                                     |

Reviewer: ```suggestion
| Async.Parallel          | Composes a new computation from multiple computations, `Async<'T> seq`, and runs them in parallel; it returns all the results in an array `Async<'T[]>`                                                                     |
```

For all functions that compose a new computation from children, if any child computations raise an exception, then the overall computation will trigger an exception. All computations that were started will receive a cancellation request.

#### Executing

| Function                     | Description                                                                                                                  |
|------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| Async.RunSynchronously       | Runs an async computation and awaits its result.                                                                             |
| Async.StartAsTask            | Runs an async computation on the ThreadPool and wraps result in a `Task<'T>`.                                                |
| Async.StartImmediateAsTask   | Runs an async computation, starting immediately on the current operating system thread, and wraps the result in a `Task<'T>` |

> | Async.Start                  | Runs an `Async<unit>` computation on the ThreadPool.                                                                         |

Author: This is what I don't know about. If I run this with Async.Start and there's an exception, I get that exception. My understanding of what "observing" means is different from what they think it should mean.

> | Async.Start                  | Runs an `Async<unit>` computation on the ThreadPool.                                                                         |

Reviewer: The point here is that Async.Start kicks off a deamon thread that nobody/nothing will be waiting for. That's worth calling out as the average person sprinkles around `Async.RuSycnronously` and hence always sees exceptions being emitted, whereas the fact that they are silently ignored in this way might be surprising.

you only get the exception if the funcion you call to produce the Async throws - if stuff within the `async {` block throws, that's deferred to runtime

for observing I'm referring to a threadpool-run task continuation not having anyone that will see/log/proapgate the exception

and alluding to how in .NET 4 it originally would tear down the process if an unhandled exception escaped from a Task continuation (i.e. if there was no .ContinueWith registered or a try/catch.

> | Async.StartImmediate         | Runs a computation, starting immediately on the current operating system thread.                                             |

Author: ```suggestion
| Async.StartImmediate         | Runs a computation, starting immediately on the current thread and continuations completing in the ThreadPool. |
```

### Cancellation

#### .NET Tasks

> .NET Tasks require that you create and share Cancellation Tokens with all sub-Tasks.

Reviewer: ```suggestion
.NET Tasks don't have any intrinsic handling of `CancellationToken`s; you are responsible for creating and passing Cancellation Tokens down the call hierarchy to all sub-Tasks.
```

Author: ```suggestion
.NET Tasks do not have any intrinsic handling of `CancellationToken`s; you are responsible for passing `CancellationToken`s down the call hierarchy to all sub-Tasks.
```

Reviewer: yes, much better

    open System
    open System.Threading
    open System.Threading.Tasks

    let loop (token: CancellationToken) = task {
        for cnt in [ 0 .. 10 ] do
            printf $"{cnt}: And..."

>             do! Task.Delay((TimeSpan.FromSeconds 1), token)  // token is required for Task.Delay to be interrupted

Reviewer: ```suggestion
            do! Task.Delay((TimeSpan.FromSeconds 1), token)  // token is required for Task.Delay to be interruptible
```

            printfn "Done"
    }

    let cts = new CancellationTokenSource (TimeSpan.FromSeconds 3)
    let runningLoop = loop cts.Token
    try
        runningLoop.GetAwaiter().GetResult()  // (1)
    with :? TaskCanceledException -> printfn "Canceled"

Output:

    0: And...Done
    1: And...Done

>     2: And...Canceled

Reviewer: ```suggestion
    1: And...Canceled
```
changed the delay to 2s to emphasize the point that it will be cancelling mid way
3x1s and a 3s timeout is all-but a race condition
